[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

dingsl-giser opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Fix problems with pyspark in Windows:
1. Fixed datetime conversion to timestamp before 1970;
2. Fixed datetime conversion when timestamp is negative;
3. Adding a test script.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Pyspark has problems serializing pre-1970 times in Windows.

An exception occurs when executing the following code under Windows:
```python
rdd = sc.parallelize([('a', datetime(1957, 1, 9, 0, 0)),
                      ('b', datetime(2014, 1, 27, 0, 0))])
df = spark.createDataFrame(rdd, ["id", "date"])

df.show()
df.printSchema()

print(df.collect())
```

```bash
  File "...\spark\python\lib\pyspark.zip\pyspark\sql\types.py", line 195, in toInternal
    else time.mktime(dt.timetuple()))
OverflowError: mktime argument out of range

at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:503)
	at org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:638)
	at org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:621)
	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:456)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:489)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:729)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:340)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:872)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:872)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:349)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:313)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:127)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	... 1 more
```
and
```bash
File ...\spark\python\lib\pyspark.zip\pyspark\sql\types.py, in fromInternal:
Line 207:   return datetime.datetime.fromtimestamp(ts // 1000000).replace(microsecond=ts % 1000000)

OSError: [Errno 22] Invalid argument
```
 After updating the code, the above code was run successfully！
```bash
+---+-------------------+
| id|               date|
+---+-------------------+
|  a|1957-01-08 16:00:00|
|  b|2014-01-26 16:00:00|
+---+-------------------+

root
 |-- id: string (nullable = true)
 |-- date: timestamp (nullable = true)

[Row(id='a', date=datetime.datetime(1957, 1, 9, 0, 0)), Row(id='b', date=datetime.datetime(2014, 1, 27, 0, 0))] 
```
### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
New and existing test suites

dingsl-giser committed:
[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

dingsl-giser committed:
[SPARK-39176][PYSPARK] Pre - 1970 time serialization test

[SPARK-39176][PYSPARK] Pre - 1970 time serialization test

dingsl-giser committed:
[SPARK-39176][PYSPARK] Pre - 1970 time serialization test

[SPARK-39176][PYSPARK] Pre - 1970 time serialization test

AnywalkerGiser commented:
@HyukjinKwon I closed the RP in the 3.0 branch(https://github.com/apache/spark/pull/36537) and raised a new RP in the master branch.

dingsl-giser committed:
[SPARK-39176][PYSPARK] Pre - 1970 time serialization test

[SPARK-39176][PYSPARK] Pre - 1970 time serialization test

dingsl-giser committed:
[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

dingsl-giser committed:
[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

HyukjinKwon commented on python/pyspark/sql/types.py:
```diff
@@ -212,15 +213,29 @@ def needConversion(self) -> bool:
-            seconds = (
-                calendar.timegm(dt.utctimetuple()) if dt.tzinfo else time.mktime(dt.timetuple())
-            )
+            if platform.system().lower() == 'windows':
```
Should we also check if the value is negative?

AnywalkerGiser replied:
Don't need, this method is still available when the negative.

HyukjinKwon replied:
How about performance? This is hot path that's executed every value, and the performance here is pretty critical.

AnywalkerGiser replied:
Are there professional performance testers?

HyukjinKwon commented:
@AnywalkerGiser BTW, it would be easier to follow and review if you have any link on the OS difference in that Python library.

AnywalkerGiser commented:
@HyukjinKwon Do you mean the platform library, added to the code comments?

dingsl-giser committed:
Merge branch 'apache:master' into master

HyukjinKwon commented:
I am surprised that the same Python library returns a different value depending on OS. I know some libraries are dependent on C library implementation but didn't expect such drastic difference. Code comment or documentation anything is fine. I would like to see if this is an official difference in Python.

AnywalkerGiser commented:
@HyukjinKwon [Python3 datetime interface documentation](https://docs.python.org/3/library/datetime.html)
This is explained in this document.
![image](https://user-images.githubusercontent.com/97273184/168725090-9f9b5320-50d2-46d0-acb0-8bb4a73d2694.png)

HyukjinKwon commented:
So is C localtime function has a different behaviour in OS? that returns negative values?

AnywalkerGiser commented:
The localtime function does not differ in OS, the fromtimestamp function does, and datetime many functions have problems resolving dates before 1970 in windows.

AnywalkerGiser commented:
@HyukjinKwon Can this solution be merged into master?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
I would like to understand the problem first, and see if this behaviour difference is internal or official. This is core code path so would have to expect more reviews and time.

dingsl-giser committed:
Merge branch 'apache:master' into master

dingsl-giser committed:
[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

dingsl-giser committed:
[SPARK-39176][PYSPARK] Pre - 1970 time serialization test

[SPARK-39176][PYSPARK] Pre - 1970 time serialization test

AnywalkerGiser commented:
Understood, looking forward to your reply later.

HyukjinKwon commented:
@AnywalkerGiser mind pointing out any documentation that states this negative value?

AnywalkerGiser commented:
Here are some blogs on related issues:
[Python | mktime overflow error](https://stackoverflow.com/questions/2518706/python-mktime-overflow-error)
[Python fromtimestamp OSError](https://stackoverflow.com/questions/37494983/python-fromtimestamp-oserror)

AnywalkerGiser commented:
I think the Spark project can look for a solution if Python doesn't fix this bug.

HyukjinKwon commented:
Does that happen in all Windows with all Python versions?

HyukjinKwon commented:
is this a bug in Python?

AnywalkerGiser commented:
I thought it was a bug in python, but the documentation said it would report an error if it was out of time range. I tested python 3.6, 3.7, and 3.8 on windows.

dingsl-giser committed:
Merge branch 'apache:master' into master

dingsl-giser committed:
[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

srowen commented on python/pyspark/sql/types.py:
```diff
@@ -212,15 +213,32 @@ def needConversion(self) -> bool:
-            seconds = (
-                calendar.timegm(dt.utctimetuple()) if dt.tzinfo else time.mktime(dt.timetuple())
-            )
+            if platform.system().lower() == "windows":
```
I'm just super surprised this is Windows specific?

AnywalkerGiser replied:
Yes, and I want you to check it out for yourself.

dingsl-giser committed:
Merge branch 'apache:master' into master

dingsl-giser commented:
@HyukjinKwon This problem still exists in the new version. Can it be merged?
